### ISSUE TITLE
[v1.1] immutable option

### ISSUE THREAD

**Issue by Yomguithereal** [Yomguithereal/baobab#202]

**Comment by Yomguithereal** [comment]
@christianalfoni, in your experience with deep freezing of object, how did you deal with arrays? Did you iterate over every items of the array to freeze them?

**Comment by christianalfoni** [comment]
Hi @Yomguithereal ,

With immutable-store I traverse into all objects and arrays and "on the way back" I freeze them. Did that make sense? :-)

**Comment by Yomguithereal** [comment]
I created an `immutable` branch on the repo. You can try an immutable Baobab tree now by doing:

```js
var tree = new Baobab({}, {immutable: true});
```

There is still one question that haunts me: do you deep clone data given to tree before freezing or do you freeze the given references?

**Comment by Yomguithereal** [comment]
*Example*

```js
var newData = {hello: 'world'};
cursor.set(newData);

// is 'newData' frozen now or should I deep clone it, then freeze my internal clone?
```

**Comment by christianalfoni** [comment]
Hi @Yomguithereal ,

In my opinion you should freeze the data passed in or the instantiated tree would not be immutable... which I think could be confusing?

Awesome stuff, I will try it out :-)

**Comment by Yomguithereal** [comment]
The question isn't to freeze or not freeze the data passed through the tree but rather to decide whether I should freeze the reference given by the user or clone the reference to create my own in the tree then freeze this internal reference.

**Comment by christianalfoni** [comment]
Ah, sorry, I misunderstood. In immutable-store I just freeze, hm... not sure if there is a side-effect to not cloning the initial data. Have not gotten any feedback on that in immutable-store though, so based on that I think it is okay *not* to clone the initial data.

**Comment by Yomguithereal** [comment]
Released a `baobab@1.1.0-immutable` version on npm for convenience.
